# TALLER Clase 25 - Modelos lineales generalizados

Alejandro Mantilla Redondo - 201711304

Ximena Palacio Argüello - 201730995

En el paquete ISLR de R, la base de datos `OJ` contiene información sobre la venta de jugos de naranja (citrus hill `CH` o Minute Maid `MM`), en función de varias variables. Use las variables: precio de CH, precio de MM, descuento en CH, desceunto en MM, y las dos indicadoras de si son productos especiales.

In [1]:
library(ISLR)
datos <- ISLR::OJ[, c("Purchase",
                      "PriceCH",
                      "PriceMM",
                      "DiscCH",
                      "DiscMM",
                      "SpecialCH",
                      "SpecialMM")]

datos$Purchase <- ifelse(datos$Purchase == "CH", 1, 0)

head(datos)


Purchase PriceCH PriceMM DiscCH DiscMM SpecialCH SpecialMM
1 1        1.75    1.99    0.00   0.0    0         0        
2 1        1.75    1.99    0.00   0.3    0         1        
3 1        1.86    2.09    0.17   0.0    0         0        
4 0        1.69    1.69    0.00   0.0    0         0        
5 1        1.69    1.69    0.00   0.0    0         0        
6 1        1.69    1.99    0.00   0.0    0         1

## 1. Estime el modelo.

$$
\ln\left(\frac{p}{1-p}\right) = \beta_0 + \beta_{\text{priceCH}}x_{\text{priceCH}} + \beta_{\text{priceMM}}x_{\text{priceMM}} + \beta_{\text{discCH}}x_{\text{discCH}} + \beta_{\text{discMM}}x_{\text{discMM}}+ \lambda_{\text{specialCH}}x_{\text{specialCH}}+ \lambda_{\text{specialMM}}x_{\text{specialMM}}
$$

Se toma como una evento positivo la compra de `CH`. $p$ es por lo tanto la probabilidad estimada de que un registro resulte en la compra de `CH`.

In [2]:
modelo <- glm(Purchase ~.,
              data = datos, family = 'binomial')
summary(modelo)



Call:
glm(formula = Purchase ~ ., family = "binomial", data = datos)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2208  -1.1650   0.7689   0.9473   1.7500  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -1.7940     1.2818  -1.400  0.16165    
PriceCH      -2.4158     0.8602  -2.809  0.00498 ** 
PriceMM       3.2613     0.6220   5.243 1.58e-07 ***
DiscCH        3.5402     0.8306   4.262 2.03e-05 ***
DiscMM       -1.2438     0.3816  -3.259  0.00112 ** 
SpecialCH     0.4081     0.2509   1.627  0.10380    
SpecialMM    -0.4951     0.2094  -2.364  0.01808 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1430.9  on 1069  degrees of freedom
Residual deviance: 1318.8  on 1063  degrees of freedom
AIC: 1332.8

Number of Fisher Scoring iterations: 4


## 2. Interprete los parámetros.

### Interceptos

In [3]:
betas <- modelo$coefficients

p = exp(betas["(Intercept)"]) / (1 + exp(betas["(Intercept)"]))
p

(Intercept) 
  0.1425822

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (tanto para `CH` como para `MM`), el precio de venta es cero y no hay descuentos, es $-1.7940$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.143$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [4]:
p = (exp(betas["(Intercept)"] + betas["SpecialCH"])
     / (1 + exp(betas["(Intercept)"] + betas["SpecialCH"])))
p

(Intercept) 
  0.2000659

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (solo para `MM`), el precio de venta es cero y no hay descuentos, es $-1.386$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.200$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [5]:
p = (exp(betas["(Intercept)"] + betas["SpecialMM"])
     / (1 + exp(betas["(Intercept)"] + betas["SpecialMM"])))
p

(Intercept) 
 0.09202636

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (solo para `CH`), el precio de venta es cero y no hay descuentos, es $-2.289$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.092$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [6]:
p = (exp(betas["(Intercept)"]
         + betas["SpecialMM"]
         + betas["SpecialCH"])
     / (1 + exp(betas["(Intercept)"]
                + betas["SpecialMM"]
                + betas["SpecialCH"])))
p

(Intercept) 
  0.1322722

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando es una venta especial (tanto para `CH` como para `MM`), el precio de venta es cero y no hay descuentos, es $-1.881$. En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.132$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

### Efectos

In [7]:
exp(betas["PriceCH"])

PriceCH 
0.08929727

Si el precio de `CH` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{precioCH}}$ unidades, es decir, $-2.416$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{precioCH}}} = 0.089$. El _odds ratio_ disminuye con incrementos en el precio de `CH`.

In [8]:
exp(betas["PriceMM"])

PriceMM 
26.08288

Si el precio de `MM` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{precioMM}}$ unidades, es decir, $3.2613$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{precioMM}}} = 26.082$. El _odds ratio_ incrementa con incrementos en el precio de `MM`.

In [9]:
exp(betas["DiscCH"])

DiscCH 
34.47444

Si el descuento de `CH` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{discCH}}$ unidades, es decir, $3.5402$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{discCH}}} = 34.474$. El _odds ratio_ incrementa con incrementos en el descuento de `CH`.

In [10]:
exp(betas["DiscMM"])

DiscMM 
0.2882987

Si el descuento de `MM` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{discMM}}$ unidades, es decir, $-1.244$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{discMM}}} = 0.288$. El _odds ratio_ disminuye con incrementos en el descuento de `MM`.

## 3. ¿Es el modelo significativo?

Relizamos una prueba de significancia global por LRT, donde la hipótesis del modelo será la siguiente:

\begin{align*}
    H_{0} &: \text{El modelo no es significativo.}\\
    H_{a} &: \neg H_{0}.
\end{align*}

In [11]:
ep <- summary(modelo)$null.deviance - summary(modelo)$deviance
p_value <- 1 - pchisq(ep, df = 7 - 1)
p_value

[1] 0

Dado que el $p$-value es menor a nuestro nivel de significancia, es decir 0 < 5%, entonces se puede rechaza la $H_0$ por lo que se puede concluir que el modelo es significativo.

## 4. ¿Es cierto que las variables de si el producto es especial no afectan las ventas?

Para comprobar si las variables que indican si el producto es especial no afectan a si la venta es `CH`. Se realiza una prueba en donde se compara con el modelo reducido sin estas variables contra el modelo completo.

$$
H_0: \lambda_{\text{specialMM}} = \lambda_{\text{specialCH}} = 0
$$

$$
H_a: \neg H_0
$$

In [12]:
modelo_red <- glm(Purchase ~ . - SpecialCH - SpecialMM, data = datos, family = "binomial")
anova(modelo_red, modelo, test = "LRT")

Resid. Df Resid. Dev Df Deviance Pr(>Chi)   
1 1065      1330.00    NA       NA          NA
2 1063      1318.76     2 11.24044 0.003623839

A partir de la prueba anova que nos arroja el $p$-value, se puede concluir que bajo un nivel de confianza del 95% las variables de si el producto es especial no afectan el modelo, dado que $p$-value = 0.00362 < 5%.

## 5. ¿Es cierto que el efecto del precio de `CH` es el negativo del efecto del precio de `MM`?

Para comprobar si el efecto del precio de `CH` es el negativo del efecto del precio de `MM`. Se realiza una prueba de Wald con el siguiente contraste de hipótesis:

$$
H_0:\beta_{\text{priceCH}} = - \beta_{\text{priceMM}}
$$

$$
H_a: \neg H_0
$$

In [13]:
x <- as.matrix(cbind(1, datos[, -1]))
var_beta <- solve(t(x)
                  %*% (diag(modelo$fitted.values
                            * (1 - modelo$fitted.values)))
                  %*% x)

a <- c(0, 1, 1, 0, 0, 0, 0)
theta <- t(modelo$coefficients) %*% a
var_theta <- t(a) %*% var_beta %*% a

ep <- theta / sqrt(var_theta)
p_value <- 1 - pchisq(ep, df = 1)
p_value

[,1]     
[1,] 0.2631522

A partir de la prueba de Wald se puede concluir bajo un nivel de confianza del 95% que el efecto del precio de `CH` no es el negativo del efecto del precio de `MM`, dado que $p$-value = 0.26315 > 5%.

## 6. Haga un intervalo para la probabilidad de compra de MM si precio CH=1.5, precio MM=2.1, descuento CH=0, descuento MM=.2, Special CH=SpecialMM=0. 

Se procede a calcular el intervalo de la siguiente manera: 

In [14]:
a2 <- c(1, 1.5, 2.1, 0, 0.2, 0, 0)
mu <- t(a2) %*% modelo$coefficients

sup <- mu + qnorm(0.975) * sqrt(t(a2) %*% var_beta %*% a2)
inf <- mu - qnorm(0.975) * sqrt(t(a2) %*% var_beta %*% a2)
supp <- exp(sup) / (1 + exp(sup))
infpp <- exp(inf) / (1 + exp(inf))
interp <- c(infpp, supp)
interp

[1] 0.6231492 0.8654796

Se obtuvo que el intervalo para la probabilidad de compra según los datos dados sería: 

$$
IP = [0.6231, 0.8654]
$$
